### ISSUE TITLE
Change (or add) from hardcoded string host OS to an Enum.

### ISSUE THREAD

**Issue by PanTrakX** [godotengine/godot-proposals#2823]
### Describe the project you are working on
--
### Describe the problem or limitation you are having in your project
Cross-platform game/app development nowadays is a highly requested feature on modern game engines.
I was looking at the Godot capabilities for that and stumbled across the ``OS.get_name()`` which returns a String.
From the docs:
> Returns the name of the host OS. Possible values are: "Android", "Haiku", "iOS", "HTML5", "OSX", "Server", "Windows", "UWP", "X11".

Hardcoded strings are bad in many ways.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation
Instead of having hardcoded strings my proposal would be to add a Platform Enum.

With this change, instead of:
```
if OS.get_name() == "Android":
    do_smth_on_android()
elif OS.get_name() == "iOS":
    do_smth_on_ios()
else:
    do_smth_generic()
```
We could do:
```
match OS.get_platform():
    OS.Platform.Android:
        do_smth_on_android()
    OS.Platform.iOS:
        do_smth_on_ios()
    _:
        do_smth_generic()
```
Which looks much more cleaner, is more secure(misspelling will not run) and probably more efficient.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams
I have including a block of code above

### If this enhancement will not be used often, can it be worked around with a few lines of script?
I guess it could.

### Is there a reason why this should be core and not an add-on in the asset library?
It is core and it will help "fight" the string hardcoding.

**Comment by Calinou** [comment]
As far as I know, platform ports can be added at will without modifying the rest of Godot's source code (especially the OS singleton). This means we need a way to return either a string or an enum value.

Also, some people may still want to have a string returned (for instance, to display it in a game's main menu or in automatically generated bug reports).

**Comment by dsnopek** [comment]
> As far as I know, platform ports can be added at will without modifying the rest of Godot's source code (especially the OS singleton

Yeah, I have at least one game that also runs on a Raspberry PI Zero via [FRT](https://github.com/efornara/frt), and my game can change how it works when running on FRT by having code like this:

```
if OS.get_name() == "FRT":
    # Use less particles or something.
    pass
```

This allows me to have the same source code for all exports of my game, with just a few conditions here and there to modify behavior to accomidate the lower specs of a Raspberry PI. If these were enums, then Godot would actually give an error if I tried to use an enum that didn't exist when running on other platforms, and so I'd have to keep the FRT code in a separate file or something.

Also, the OS names are used in project settings, for having settings overrides for a particular platform, so we'd need to still have the strings for that as well.

**Comment by PanTrakX** [comment]
So it looks like ''get_name()'' doesn't return hardcoded strings so my proposal would not work. Then how does it work? I looked on the source code under core/os/os.cpp but couldn't find the implementation there.

**Comment by katoneko** [comment]
`OS.get_name()` implies that it returns the _name_ of the platform and not some object (an enum) to represent it. To illustrate the difference: `get_os()` and `get_os_name()` are not the same. So, I'd agree with Calinou: keep `OS.get_name()` as is but (optionally) add another function to return an enum to reduce the risk of typos when checking against the name string.

**Comment by Nikitf777** [comment]
Maybe it's better to just expose OS's derived classes to the public API, so you can use the `is` operator to check what OS you are on (like `if OS.get_singleton() is OS_Android:`)? It also makes easier to add some OS-specific methods to Godot. Now, some method of the OS class work not on all platforms. With this change, these methods will be accessible only from a specific OS-derived class, like `OS_Android`. I'm thinking of creating a separate proposal for it describe it more in details.
